### PR TITLE
Update Alpaca and Polygon API for new v2 endpoints

### PR DESCRIPTION
**Pull request by Petersoj**
This PR updates all the Alpaca and Polygon endpoints along with adding optimizations and organizations to `alpaca-java`.

~@mainstringargs There seems to be a lot of dynamically generated JSON objects that would come from various responses in the [/v2/account/activities](https://docs.alpaca.markets/api-documentation/api-v2/account-activities/) endpoint because there is a lot of various activity updates that can happen to an account as shown [here](https://docs.alpaca.markets/api-documentation/api-v2/account-activities/#activity-types). Any ideas on how we should implement such a dynamic endpoint with static JSON schema sources? I'm thinking that we just pass on the `JsonObject` and the user of this library can get the data they need themselves. That would be the easiest, but not so elegant, approach. Let me know what you think.~

### CODE REVIEW

**Comment by mainstringargs** [comment]
Interesting -- I hadn't looked at this yet.  I may be incorrect, but isn't there really just 2 types of json responses -- TradeActivity and NonTradeActivity?  The TradeActivity will always have an activity_type of FILL and the NonTradeActivity can have any of these as the activity_type:  https://docs.alpaca.markets/api-documentation/api-v2/account-activities/#activity-types?  What I'm saying is that each activity type doesn't really have its own information, it will just be populating a NonTradeActivity object (with some fields unset for some activity types -- like the last 2 are only relevant for dividends).  Is that correct or am I missing the boat?

**Comment by Petersoj** [comment]
@mainstringargs Yes you are correct. I was testing this endpoint yesterday and thought I saw objects with different values, but testing it again shows that the JSON returned does indeed correspond with the JSON in the docs. Depending on the activity type, the API will simply return JSON for a Non-Trade Account Activity Entity and omit the key value pairs that aren't used by that particular activity type. 
Thanks for taking a look at this!

**Comment by mainstringargs** [comment]
Hey -- is this one still in progress?  Thinking I'll push a new 5.0.0 release once this is ready to go

**Comment by Petersoj** [comment]
@mainstringargs Yes both my open PRs are in progress right now. They should be complete by the end of this weekend hopefully. I think we should update all endpoints for Polygon and Alpaca and do some testing and then release 5.0.0.

**Comment by mainstringargs** [comment]
Sounds good!

**Comment by Petersoj** [comment]
@mainstringargs Looks like the update to v2 is going to take a little longer than I though as I wasn't able to finish this weekend. I'm implementing many optimizations and organizations in addition to updating everything to v2 so there will be quite a few bugs, but those will get ironed out when extensive testing is done 👍

**Comment by mainstringargs** [comment]
No problem.  Thanks for the update!  Take your time.

**Comment by Petersoj** [comment]
@mainstringargs This is ready for review finally!
I've made a number of changes including optimizations, organizations, and updates:
- Make request, exception, and websocket classes abstract.
- Update all Alpaca and Polygon endpoints
- Make all JSON sources schemas for jsonschema2pojo (to allow for more flexibility in structure, naming, and documentation of JSON sources)
- Other refactoring and optimizations

When the market opens tomorrow, I'll do extensive testing and make sure there are no issues, then this should be good to merge. It would be great if you could review/test all this too.
Thanks!

**Comment by mainstringargs** [comment]
Great work!  I’ll review it, but won’t be for a couple days

File: build.gradle [hunk 1]
```diff
@@ -136,76 +132,49 @@ sourceSets {
...
+    file(jsonSourceDirectory).eachFileRecurse { file ->
+        if (file.getName().endsWith(".json")) {
+            def startPackageIndex = file.getAbsolutePath().indexOf(jsonSourceDirectory) + jsonSourceDirectory.length()
+            def targetPackage = jsonPackageNameStart + file.getAbsolutePath()
+                    .replace('/' + file.getName(), '')
+                    .substring(startPackageIndex)
+                    .replace('/', '.').replace('-', '').replace('_', '')
+                    .toLowerCase()
+
+            def jsonToPOJOTask = tasks.create('json2POJOTask-' + targetPackage + "." + file.getName(), GradleBuild) {
```
Comment (mainstringargs) [inline build.gradle]: I'm getting some errors when doing gradlew build -- and I'm trying to decide if it's something with my environment or a difference between Linux and Windows.   It looks like it is putting in slashes where it should be putting in periods when it is generating sources from the Json:

image

image

Have you seen this or any thoughts on what may be happening?

File: build.gradle [hunk 2]
```diff
@@ -136,76 +132,49 @@ sourceSets {
...
-
-        def jsonToPOJOTask = tasks.create('jsonToPOJOTask-' + targetPackage, GradleBuild) {
-            startParameter.projectProperties = ['source'         : jsonSourceDirectory + path,
-                                                'sourceType'     : sourceType,
-                                                'targetPackage'  : targetPackage,
-                                                'targetDirectory': targetDirectory]
-            buildFile = 'sampleJson2Pojo.gradle'
-            tasks = ['generateJsonSchema2Pojo']
-        }
+    final def jsonSourceDirectory = "${project.projectDir}/sample_json/"
```
Comment (mainstringargs) [inline build.gradle]: Love that you went to schemas.  What do you think about changing this directory name now that it is schema based?

File: sample_json/alpaca/account/account.json [hunk 3]
```diff
@@ -1,28 +1,110 @@
...
-  "short_market_value": "0",
-  "shorting_enabled": true,
-  "sma": "0",
-  "status": "ACTIVE",
-  "trade_suspended_by_user": false,
-  "trading_blocked": false,
-  "transfers_blocked": false
-}
\ No newline at end of file
+  "type": "object",
```
Comment (mainstringargs) [inline sample_json/alpaca/account/account.json]: For future reference -- how were these schemas created?  Is it basically just looking at the alpaca documentation and creating these files?  Or is there another step?

File: build.gradle (hunk 1 above)
Comment (Petersoj) [reply build.gradle]: Ahh I see what's happening. I develop on Mac so I neglected to replace the Windows file separator character when creating a unique task name. I'll fix this later today.

File: build.gradle (hunk 2 above)
Comment (Petersoj) [reply build.gradle]: Schemas are great right!!! What were you thinking I should change the directory name to? Maybe: `schema_json`?

File: sample_json/alpaca/account/account.json (hunk 3 above)
Comment (Petersoj) [reply sample_json/alpaca/account/account.json]: For generating the schemas, I created a simple DOM parser using XPaths which allow you to grab XML tags really easily without digging deep through a bunch of nodes and such. So you can just copy and paste in Alpaca or Polygon Documentation HTML and automatically generate the schemas (adding a few adjustments as needed, such as inheritance). The util class is located [here](https://github.com/mainstringargs/alpaca-java/pull/21/files#diff-e04eea01e5149706e1ef7e5cce402d4e).

File: build.gradle (hunk 2 above)
Comment (mainstringargs) [reply build.gradle]: Yes! schema_json would work

File: sample_json/alpaca/account/account.json (hunk 3 above)
Comment (mainstringargs) [reply sample_json/alpaca/account/account.json]: Very slick

File: build.gradle (hunk 1 above)
Comment (Petersoj) [reply build.gradle]: Fixed! I've simply replaced `/` with `File.separator`.

**Comment by Petersoj** [comment]
I've tested many things on my end and it all is working great! This should be good to merge and integrated into the new version release. With schemas and the abstraction implemented, it'll be easier to implement new API changes with Alpaca and Polygon so we can keep this library up-to-date and consistent.

**Comment by mainstringargs** [comment]
I pulled in the latest, and I'm seeing this:
image

**Comment by Petersoj** [comment]
@mainstringargs Try my new push. Sorry for all the debugging. I don't have access to a Windows machine...

**Comment by mainstringargs** [comment]
Getting warmer -- now I'm getting another error:

image

Thinking its a cloning issue?  Or are you seeing this too?

**Comment by Petersoj** [comment]
My build has been building successfully this entire time with no warnings haha. This is a file encoding issue. All the file encodings are in UTF-8, but it looks like your computer's default encoding for the gradle javadoc plugin is something else (looks like Cp1252 which I've never heard of). I've explicitly set the encoding for the javadoc plugin in my latest push. Try it now with `gradlew clean build`. Again, sorry for the debugging! I appreciate your fast responses!

**Comment by mainstringargs** [comment]
Ah!  There we go... Built this time!   I'll take a closer look now.  Thanks!!

**Comment by Petersoj** [comment]
Hey don’t merge this quite yet! I’ve found a bug with order requesting. Will fix soon!

**Comment by mainstringargs** [comment]
No problem.  I’ll review a bit more, then I’ll merge tomorrow unless you
need me to hold it for any reason

**Comment by Petersoj** [comment]
@mainstringargs I've fixed the problem with submitting market orders (a precondition check was causing that method to fail on Market orders) and I've added overload methods for the various types of order requesting for parameter simplicity (e.g. might be confusing for users who have an OrderType.MARKET to have limitPrice and stopPrice in the parameter list).
Everything looks good to be merged (hopefully)! Remember to update the javadoc link and the version in the usage section of the README. Thanks!
After this big merge, I think bug fixes should be rolled out more regularly (as in uploaded to maven central ASAP) as there is a surprising number of individuals who use this library :)

**Comment by mainstringargs** [comment]
Cool.  Great job!  I will do all of that and push up a 5.0.0 release
tomorrow after I test it out once the market opens up.

**Pull request merged**